libraries/Microsoft.Bot.Builder.Dialogs/Memory/DialogStateManager.cs

```diff
@@ -375,6 +377,31 @@ public async Task SaveAllChangesAsync(CancellationToken cancellationToken = defa
             }
         }
 
+        /// <summary>
+        /// Save all changes with a trace snapshot activity (only if different from last trace).
+        /// </summary>
+        /// <param name="traceLabel">trace label for the memory snapshot.</param>
+        /// <param name="cancellationToken">cancellationtoken.</param>
+        /// <returns>task.</returns>
+        public async Task SaveAllChangesAsync(string traceLabel, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // save changes
+            await SaveAllChangesAsync(cancellationToken).ConfigureAwait(false);
+
+            // trace memory snapshot
+            var snapshot = GetActiveDialogContext(this.dialogContext).State.GetMemorySnapshot();
+
+            // only trace if it's changed since last time we traced it.
+            var snapshotHash = StringUtils.Hash(JsonConvert.SerializeObject(snapshot));
+            this.dialogContext.Context.TurnState.TryGetValue("lastSnapshotHash", out object lastSnapshotHash);
```

Comment:
Should we namespace "lastSnapshotHash" to avoid potential collisions in turn state? 
Something like "Microsoft.Bot.Builder.Dialogs.Memory.DialogStateManager.lastSnapshotHash"
Also consider making this string a private constant.